src/main/java/com/rackspace/salus/acm/services/AgentInstallService.java

```
@@ -0,0 +1,459 @@
+  private final BoundEventSender boundEventSender;
+
+  @Autowired
+  public AgentInstallService(JdbcTemplate jdbcTemplate,
+                             EntityManager entityManager,
+                             AgentReleaseRepository agentReleaseRepository,
+                             AgentInstallRepository agentInstallRepository,
+                             BoundAgentInstallRepository boundAgentInstallRepository,
+                             ResourceApi resourceApi,
+                             BoundEventSender boundEventSender) {
+    this.jdbcTemplate = jdbcTemplate;
+    this.em = entityManager;
+    this.agentReleaseRepository = agentReleaseRepository;
+    this.agentInstallRepository = agentInstallRepository;
+    this.boundAgentInstallRepository = boundAgentInstallRepository;
+    this.resourceApi = resourceApi;
+    this.boundEventSender = boundEventSender;
+  }
+
+  public AgentInstallDTO install(String tenantId, AgentInstallCreate in) {
+    Assert.notNull(tenantId, "tenantId is required");
+
+    log.debug("Creating install={} for tenant={}", in, tenantId);
+
+    final AgentRelease agentRelease = agentReleaseRepository.findById(in.getAgentReleaseId())
+        .orElseThrow(() -> new NotFoundException("Could not find associated agent release"));
+
+    final List<AgentInstall> existing = agentInstallRepository
+        .findAllByTenantIdAndAgentRelease_Id(tenantId, in.getAgentReleaseId());
+
+    if (existing.stream()
+        .anyMatch(agentInstall -> agentInstall.getLabelSelector().equals(in.getLabelSelector()))) {
+      throw new AlreadyExistsException("AgentInstall with same release and label selector exists");
+    }
+
+    final AgentInstall agentInstall = new AgentInstall()
+        .setAgentRelease(agentRelease)
+        .setLabelSelector(in.getLabelSelector())
+        .setTenantId(tenantId);
+
+    final AgentInstall saved = agentInstallRepository.save(agentInstall);
+
+    bindInstallToResources(saved);
```

i'm a little concerned about error handling.  If there are networking problems and the resourceApi cant be reached, this bindInstallToResources() call will fail silently and the user will think everything is ok.  Shouldn't we be more careful about stuff like that?

---

or possibly it will get a runtime exception.  but then if the user calls install again, they will get the AlreadyExistsException

---

As @jjbuchan mentioned, we have this same problem elsewhere. The risk is minimized by the multiple replicas we're running of each of these backend services. So the cluster would have to be in a very bad state for one of these inter-API calls to fail.

In any case, I'm using Ambassador as my trial area for Spring Retry, such as

https://github.com/racker/salus-telemetry-ambassador/blob/a7aeba959e7f1a9d304f26b9d4545dbd9b663fb0/src/main/java/com/rackspace/salus/telemetry/ambassador/services/AgentInstallsListener.java#L98